### PR TITLE
docker_swarm_service_info: Read information about swarm services

### PR DESCRIPTION
**Pull request by hannseman**
##### SUMMARY

This PR adds a new module for reading information from docker swarm services. It reflects the `docker service inspect` command in CLI and `GET /services/{id}` in the Docker API.

I followed the structure of previous docker-related `*_info` modules and added two methods for inspecting services in `AnsibleDockerSwarmClient`. 

Not really sure why the `json.dumps` and `json.loads` are used on the return value from docker-py, I added it to this implementation as well as it's used throughout `AnsibleDockerSwarmClient` but I'm quite interested in why it's needed.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

docker_swarm_service_info

### CODE REVIEW

**Comment by felixfontein** [comment]
> This PR adds a new module for reading information from docker swarm services. It reflects the `docker service inspect` command in CLI and `GET /services/{id}` in the Docker API.

One could also ask: why doesn't it return the same information as `docker_swarm_service` does? I'm not sure what's better, or if it would be helpful to have both.

> I followed the structure of previous docker-related `*_info` modules

That's only what *some* of the `_info` modules do. I guess one big question is whether users of `docker_swarm_service_info` actually want to be able to get information on various services, or whether they are interested in one service.

Actually, listing all service inspect results can already be done with the `docker_swarm_info` module. And getting the inspect result for a list of services could be done with a `loop` construct if `docker_swarm_service_info` would behave more like `docker_container_info`, `docker_network_info` and `docker_volume_info`, i.e. it gives information on one object at a time.

> Not really sure why the `json.dumps` and `json.loads` are used on the return value from docker-py, I added it to this implementation as well as it's used throughout `AnsibleDockerSwarmClient` but I'm quite interested in why it's needed.

That's actually a very good question. I've missed that so far... @tbouvet do you know why they are there? This seems to come from the initial `docker_swarm` commit (in #38280, it's already in the very first commit of that PR).

**Comment by hannseman** [comment]
> One could also ask: why doesn't it return the same information as docker_swarm_service does? I'm not sure what's better, or if it would be helpful to have both.

Yes I also though about passing the result of `inspect_service` to `docker_swarm_service.DockerService` and return the result as `DockerService.get_facts` . It felt best to just return the raw result as the `get_facts` currently returns the more flat service representation which is deprecated in `docker_swarm_service`. Like `restart_policy_delay` vs `restart_config.delay`.

Documentation is also a lot easier to maintain as we can rely on the Docker API documentation. 

When just returning the raw result of ` inspect_service` we can also return a few parameter that we are not returning in `docker_swarm_service`, like `Version` and `CreatedAt` / `UpdatedAt`.


>That's only what some of the _info modules do. I guess one big question is whether users of docker_swarm_service_info actually want to be able to get information on various services, or whether they are interested in one service.

>Actually, listing all service inspect results can already be done with the docker_swarm_info module. And getting the inspect result for a list of services could be done with a loop construct if docker_swarm_service_info would behave more like docker_container_info, docker_network_info and docker_volume_info, i.e. it gives information on one object at a time.

Ah I actually missed that docker_swarm_info could do this. I think it makes sense to only return a single service especially if this feature already exists in docker_swarm_info. I'll update it to only work with single services.

**Comment by felixfontein** [comment]
> > One could also ask: why doesn't it return the same information as docker_swarm_service does? I'm not sure what's better, or if it would be helpful to have both.
> 
> Yes I also though about passing the result of `inspect_service` to `docker_swarm_service.DockerService` and return the result as `DockerService.get_facts` . It felt best to just return the raw result as the `get_facts` currently returns the more flat service representation which is deprecated in `docker_swarm_service`. Like `restart_policy_delay` vs `restart_config.delay`.

Hmm, since when is the flat representation deprecated? I don't mind deprecating it, but we should then say so in the docs (and it needs a porting guide entry). And it would be nice if an alternative is returned (f.ex. the inspect result).

> Documentation is also a lot easier to maintain as we can rely on the Docker API documentation.

That's definitely true!

> Ah I actually missed that docker_swarm_info could do this. I think it makes sense to only return a single service especially if this feature already exists in docker_swarm_info. I'll update it to only work with single services.

:thumbsup:

**Comment by hannseman** [comment]
> Hmm, since when is the flat representation deprecated? I don't mind deprecating it, but we should then say so in the docs (and it needs a porting guide entry). And it would be nice if an alternative is returned (f.ex. the inspect result).

You're right. We never deprecated the return values when we deprecated the "flat parameters" in preference of the dict versions (ex `restart_policy_*` vs `restart_config`). Just now saw that we don't even return the new dict versions.  We should probably fix this.

The problem with returning an `inspect` result in docker_swarm_service is that the update is made async so we don't really know when `inspect` will return the updated values. Today we return the parameters that was used to build the API request.

**Comment by felixfontein** [comment]
> You're right. We never deprecated the return values when we deprecated the "flat parameters" in preference of the dict versions (ex `restart_policy_*` vs `restart_config`). Just now saw that we don't even return the new dict versions. We should probably fix this.

Good question. The problem with changing return variables is that there is no good way of informing users about it, except mentioning it in porting guide and changelog entries, which some users tend to ignore. And if we change it, we should change it only for a new major version, i.e. either we change it until Thursday, or we keep it (until maybe Ansible 2.9) :)

> The problem with returning an `inspect` result in docker_swarm_service is that the update is made async so we don't really know when `inspect` will return the updated values. Today we return the parameters that was used to build the API request.

Hmm, yes, I guess we shouldn't return inspect results then :)

**Comment by hannseman** [comment]
I've limited the module to only fetch one service. IMO I think we should stick with the current return structure especially when most other `*_info` modules return the raw docker inspect response.

**Comment by WojciechowskiPiotr** [comment]
> I've limited the module to only fetch one service. IMO I think we should stick with the current return structure especially when most other `*_info` modules return the raw docker inspect response.

I kind of disagree it should return one service info only. In my opinion, this should work as `docker_node_info` accepting a single parameter and a list.

**Comment by felixfontein** [comment]
You can always use `loop` to run the module for every service name you're interested in. Or use `docker_swarm_info` to list all services.

Is there any valid use-case which really requires a list of names to be input? Please note that this is not compatible to have a single return value with facts (you always need array notation), there's no `exist` return value (you have to check array length instead), ... There should be really good reasons why the interface should get more complicated for the "I want infos on one service" use-case.

**Comment by WojciechowskiPiotr** [comment]
I was more thinking about the consistency across `_info` modules. A use case is also when you want returned value in one variable

**Comment by felixfontein** [comment]
If we want 100% consistency (independent of use-cases etc.), `docker_node_info` and `docker_image_info` should be changed, not the others :) (Also, just wanting to have output in a very specific manner is not a use-case.)

**Comment by WojciechowskiPiotr** [comment]
It is use-case in a way that lack of flexibility in module increases the playbook complexity

**Comment by hannseman** [comment]
Since this is my first `_info` module and not very familiar with all the new `docker_*_info` modules  I'm not really comfortable with taking a stronger stance here. 

I think that the most common use-case of this module would be to retrieve information about a single service, for example getting the current replica count which could've been changed by some autoscaling mechanism and using that when updating the service to not override the current replica count.

I can also see use-cases where one would like to collect all service ids in a swarm but it feels like this is a "swarm"-context and not a "service"-context and it makes sense to use the `docker_swarm_info`-module which has this functionality.

**Comment by felixfontein** [comment]
> I think that the most common use-case of this module would be to retrieve information about a single service, for example getting the current replica count which could've been changed by some autoscaling mechanism and using that when updating the service to not override the current replica count. (This exact use-case is already possible to resolve by using `relicas: -1` but I think we should try and move away from that functionality in the future).

This is also what I think the main use of this module will be. Or to check whether a service is already there or not (for which having an `exists` flag is nice to have).

If you can change the module in this way, I hope we'll manage to get this merged before feature freeze tomorrow evening.

> I can also see use-cases where one would like to collect all service ids in a swarm but it feels like this is a "swarm"-context and not a "service"-context and it makes sense to use the `docker_swarm_info`-module which has this functionality.

:thumbsup:

**Comment by hannseman** [comment]
>  Or to check whether a service is already there or not (for which having an exists flag is nice to have).

Makes sense. I've added an exists boolean as a return value.

File: lib/ansible/modules/cloud/docker/docker_swarm_service_info.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+exists:
+    description:
+      - Returns whether the service exists.
+    type: bool
+    returned: always
+    sample: true
+service:
+    description:
+      - A dictionary representing the current state of the service. Matches the C(docker service inspect) output.
+      - Will be C(None) if service does not exist.
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service_info.py]:
```suggestion
      - Will be C(none) if service does not exist.
```
That thing is called `none` in jinja2. (`xxx is none` works, while `xxx is None` does not -- at least when I last tested that.)

File: test/integration/targets/docker_swarm_service_info/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- include_tasks: test_docker_swarm_service_info.yml
```
Comment (felixfontein) [inline test/integration/targets/docker_swarm_service_info/tasks/main.yml:2]:
```suggestion
---
- include_tasks: test_docker_swarm_service_info.yml
```

Comment (hannseman) [reply test/integration/targets/docker_swarm_service_info/tasks/main.yml]: Add yaml 3-dash

File: lib/ansible/modules/cloud/docker/docker_swarm_service_info.py (hunk 1 above)
Comment (hannseman) [reply lib/ansible/modules/cloud/docker/docker_swarm_service_info.py]: You're right. Always confuses me since `False` and `True` works.

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service_info.py]: It used to confuse me too, but I think I now got used to it after a (long) time ;)

File: lib/ansible/modules/cloud/docker/docker_swarm_service_info.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    service = client.module.params['name']
+    return client.get_service_inspect(
+        service_id=service,
+        skip_missing=True
+    )
+
+
+def main():
+    argument_spec = dict(
+        name=dict(type='str'),
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service_info.py]: This should be required (also need to update docs).

Comment (hannseman) [reply lib/ansible/modules/cloud/docker/docker_swarm_service_info.py]: Good catch.

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@hannseman thanks for your work on this!
@WojciechowskiPiotr thanks for discussing!
@mkrizek thanks for merging!

**Comment by hannseman** [comment]
@felixfontein @WojciechowskiPiotr thanks for the review!
@mkrizek thanks for merging!

**Pull request merged**